expression/builtin_grouping.go

```diff
@@ -0,0 +1,206 @@
+	mode          tipb.GroupingMode
+	groupingMarks map[int64]struct{}
+}
+
+func (b *builtinGroupingImplSig) SetGroupingMode(mode tipb.GroupingMode) {
+	b.mode = mode
+}
+
+func (b *builtinGroupingImplSig) SetMetaGroupingMarks(groupingMarks map[int64]struct{}) {
+	b.groupingMarks = groupingMarks
+}
+
+func (b *builtinGroupingImplSig) getGroupingMode() tipb.GroupingMode {
+	return b.mode
+}
+
+// metadata returns the metadata of grouping functions
+func (b *builtinGroupingImplSig) metadata() proto.Message {
+	args := &tipb.GroupingFunctionMetadata{}
+	*(args.Mode) = uint32(b.mode)
+	for groupingMark, _ := range b.groupingMarks {
+		args.GroupingMarks = append(args.GroupingMarks, uint64(groupingMark))
+	}
+	return args
+}
+
+func (b *builtinGroupingImplSig) Clone() builtinFunc {
+	newSig := &builtinGroupingImplSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	newSig.mode = b.mode
+	newSig.groupingMarks = b.groupingMarks
+	return newSig
+}
+
+func (b *builtinGroupingImplSig) getMetaGroupingMarks() map[int64]struct{} {
+	return b.groupingMarks
+}
+
+func (b *builtinGroupingImplSig) getMetaGroupingID() int64 {
+	var metaGroupingID int64
+	grouping_ids := b.getMetaGroupingMarks()
+	for key := range grouping_ids {
+		metaGroupingID = key
+	}
+	return metaGroupingID
+}
+
+func (b *builtinGroupingImplSig) checkMetadata() error {
+	mode := b.getGroupingMode()
+	grouping_ids := b.getMetaGroupingMarks()
+	if mode != tipb.GroupingMode_ModeBitAnd && mode != tipb.GroupingMode_ModeNumericCmp && mode != tipb.GroupingMode_ModeNumericSet {
+		return errors.Errorf("Mode of meta data in grouping function is invalid. input mode: %d", mode)
+	} else if (mode == 1 || mode == 2) && len(grouping_ids) != 1 {
+		return errors.Errorf("Invalid number of groupingID. mode: %d, number of groupingID: %d", mode, len(b.groupingMarks))
+	}
+	return nil
+}
+
+func (b *builtinGroupingImplSig) groupingImplBitAnd(groupingID int64, metaGroupingID int64) int64 {
+	if groupingID&metaGroupingID > 0 {
+		return 1
+	}
+	return 0
+}
+
+func (b *builtinGroupingImplSig) groupingImplNumericCmp(groupingID int64, metaGroupingID int64) int64 {
+	if groupingID > metaGroupingID {
+		return 1
+	}
+	return 0
+}
+
+func (b *builtinGroupingImplSig) groupingImplNumericSet(groupingID int64) int64 {
+	grouping_ids := b.getMetaGroupingMarks()
+	_, ok := grouping_ids[groupingID]
+	if ok {
+		return 0
+	}
+	return 1
+}
+
+func (b *builtinGroupingImplSig) grouping(groupingID int64) int64 {
+	switch b.mode {
+	case tipb.GroupingMode_ModeBitAnd:
+		return b.groupingImplBitAnd(groupingID, b.getMetaGroupingID())
+	case tipb.GroupingMode_ModeNumericCmp:
+		return b.groupingImplNumericCmp(groupingID, b.getMetaGroupingID())
+	case tipb.GroupingMode_ModeNumericSet:
+		return b.groupingImplNumericSet(groupingID)
+	}
+	return 0
+}
+
+// evalInt evals a builtinGroupingSig.
+func (b *builtinGroupingImplSig) evalInt(row chunk.Row) (int64, bool, error) {
+	err := b.checkMetadata()
```

Comment:
It looks like very inefficient to call `checkMetadata` for every row, how about this:
1. merge `SetGroupingMode` and `SetMetaGroupingMarks` into one function like `SetMetadata`, and only call `checkMetadata` once in `SetMetadata`
2. add a flag `metaInited`, and set it to true in `SetMetadata`
3. only check `metaInited` during runtime

Reply:
> 

okk, I think it's a good idea.